File: golang-devildyw/middle/L114二叉树展开为链表.go
package main

func main() {}

func flatten(root *TreeNode) {
	curr := root
	for curr != nil {
		if curr.Left != nil {
			left := curr.Left
			predecessor := left
			// 查找左子树的右子树叶子节点的位置 再把这个节点接到父节点的右节点上
			for predecessor.Right != nil {
				predecessor = predecessor.Right
			}
			// 接上去
			predecessor.Right = curr.Right
			// 左子树变成右子树了
			curr.Left, curr.Right = nil, left
		}
		// 去寻找下一个节点
		curr = curr.Right
	}
}

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L114二叉树展开为链表.java
package top.dyw.leetcode.Hot100;

public class L114二叉树展开为链表 {
    public void flatten(TreeNode root) {
        if (root == null) {
            return ;
        }

        TreeNode cur = root;

        while (cur != null) {
            if (cur.left != null) {
                TreeNode left= cur.left;
                TreeNode predecessor = left;
                while (predecessor.right != null) {
                    predecessor = predecessor.right;
                }
                //接上去
                predecessor.right = cur.right;
                cur.left = null;
                cur.right = left;
            }
            cur = cur.right;
        }
    }
}
